docs/life-of-a-request.md

```diff
@@ -0,0 +1,200 @@
+```
+zproject/urls.py
+zproject/legacy_urls.py
+```
+
+There are HTML-serving, REST API, legacy, and webhook url patterns.
+
+## Views serving HTML are internationalized by server path
+
+If you look in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py), you'll see something called
+`i18n_urls`. These urls show up in the address bar of the browser, and
+serve HTML.
+
+For example, the `/hello` page (preview [here](https://zulip.com/hello/))
+gets translated in Chinese at `zh-cn/hello/` (preview [here](https://zulip.com/zh-cn/hello/)).
+
+Note the `zh-cn` prefix--that url pattern gets added by `i18n_patterns`.
+
+## API endpoints use REST
+
+With the exception of Webhooks (which we do not control), and legacy
+endpoints, and logged-out endpoints, Zulip uses REST for its API. This
+means that we use:
+
+* POST for creating something new where you don't have a unique ID
+* PUT for creating or completely replacing something for which you have
+  a unique ID
+* DELETE for deleting something
+* PATCH for updating or editing attributes of something.
+* GET to get something (read-only)
+* HEAD to check the existence of something to GET, without getting it;
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
+* POST for everything else
+
+Of these, PUT, DELETE, HEAD, OPTIONS, and GET are *idempotent*, which
+means that you can send the request multiple times and get the same
+state on the server. POST is not idempotent--if I send a message
+multiple times, I will get multiple message. PATCH is special--it can be
+idempotent, and we like it if you write API endpoints where it is
+idempotent, as much as possible.
+
+### PUT vs POST vs PATCH
+
+The difference between these verbs is subtle at first glance, but
+important.
+
+Let's say you have something (the HTTP RFC calls them "resources") like
+a realm emoji. An emoji has two attributes: a name and a url for the
+emoji image. Let's pretend there's also an emoji alternative name
+attribute.
+
+In this case, we should avoid POST, since we have a unique identifier:
+the name. We don't want to create multiple emoji with the same name--
+otherwise it would be ambiguous which emoji we want to select when
+composing a message.
+
+If we want create the emoji, we should use PUT, since we're replacing
+the entire resource at the unique ID (name). If we wanted to simply
+update some attribute of the emoji, either the image url or the
+alternative name, we should use PATCH. Otherwise, to update the image
+url without clearing the alternative name would require knowing the
+alternative name of the emoji we want to upate. If we wanted to make it
+so that you couldn't update the image url without changing the
+alternative name, or at least knowing it, we could enforce using PUT
+for updates to the realm emoji.
+
+http://restful-api-design.readthedocs.io/en/latest/methods.html can be
+a good read.
```

Comment:
Actually I remove it entirely, since it does different things with PUT than we do. We limit PUT to creation of resources, and don't use it for updating resources.